### ISSUE TITLE
Idea: a production r-multiverse (or an "r-releases" universe)

### ISSUE THREAD

**Issue by wlandau** [r-universe-org/help#365]
Following #363, I have an idea for an "r-multiverse" for production: automatically generate a release-oriented `packages.json` from an upstream list of universes. It's a somewhat different take than what I had first imagined for #363 because of the potential for name conflicts and development versions. Here is how I picture it:

1. Suppose https://github.com/r-multiverse/r-multiverse.r-universe.dev is a regular universe with a `packages.json` file, except that it also has a file called `universes.json`.
2. `universes.json` is a list of R universes. A package author can submit a PR to add a personal universe to the list. The PR triggers a GitHub Actions job to check if the universe exists, `universes.json` is still valid, no entries are duplicated, and none are removed. If the check passes, the PR automatically merges. PRs that change or remove universes will probably need manual review.
3. A GitHub Actions workflow combs through the universes in `universes.json`. For each package in each universe, the job retrieves its URL and the date of its first release.
4. A package enters `packages.json` if its name is unique across all the universes in `universe.json`. If there are name conflicts, then the package with the earliest initial release date preferred. For other name conflicts, the package is not added to `packages.json`. Instead, the name and URL of the conflicting package are printed to the GitHub Actions log, and the package author is automatically notified. 
5. Packages in  `packages.json` have `branch: "*release"` to track releases only (c.f. https://github.com/shikokuchuo/shikokuchuo.r-universe.dev/commit/27e6353489ad678823616034293b862ad9f0fbc6). (Although I'm not sure I understand exactly how `branch: "*release"` works, i.e. if there are edge cases where the release is not pulled or if a development version could supersede a release.)

I think this proposal addresses the difficult issues of name conflicts and releases. I do foresee challenges:

* It seems vitally important to run `R CMD check` and a reverse dependency check on every package release, as well as `R CMD check`  on a regular schedule. Even though only releases are tracked, is it too computationally expensive to run all these checks for such a large universe? (I wonder if it would help to put a cap on check time. I don't want to de-incentivize testing, but I do want to be mindful of resources. It would not be realistic to think that every kind of test could be run on GitHub Actions anyway.)
* What do we do if `R CMD check` or revdep checks fail? At the very least, I think the package author should be automatically notified. And I would support other types of non-disruptive incentives and nudges. I just don't want to trigger CRAN-like policies   like manual review or archival because they could de-incentivize testing.

But I think these challenges are easier than the ones that #363 started with.

**Comment by jeroen** [comment]
My main concern is not computation resources but the manual maintenance work.

`branch: "*release"` is an alias for the "latest" github release such as: https://github.com/ropensci/targets/releases/latest. You can get the same with remotes/pak like `pak::pak('ropensci/targets@latest')`. Hence this is only useful if upstream package authors diligently add git-tags to their releases, which is not generally the case. If they forget to create git tags we end up with outdated packages in the multiverse, when upstream authors release new CRAN versions

> It seems vitally important to run R CMD check and a reverse dependency check on every package release,

What do we do with this information? When I release a new version of `jsonlite` or `curl` there are always a handful revdeps with some random test failures. I know this in advance. To whom is this vitally important? To the author of the dependent package? Or the multiverse maintainers?

**Comment by wlandau** [comment]
> But your point about falling back on CRAN does give me pause. I hope we can start building this special universe from the ground up using packages whose entire dependency graphs do not need CRAN.

Maybe https://github.com/r-lib/ is a good one to start with? Maybe that one gives us the `tidyverse`. `ropensci` would be a great test case.

**Comment by shikokuchuo** [comment]
@jeroen I want to pick up on this one point. This repository for production purposes cannot rely on CRAN. The guarantee of continued availability is too low – as an example, I came across this only the other day: https://github.com/r-lib/remotes/issues/781 We want to automate as much as possible and not let arbitrary decisions get in the way.

The release tag capability is already incredibly useful, I have to thank you again for pointing this out. Indeed I have started using it, and changed my release policy to R-universe first, see for example: https://shikokuchuo.net/mirai/#installation This doesn’t mean I stop publishing on CRAN, but it means I choose when to push out releases and CRAN can catch up as and when it lets me.

**Comment by shikokuchuo** [comment]
@jeroen I want to pick up on this one point. This repository for production purposes should not rely on CRAN - the guarantee of continued availability is too low. We want to automate as much as possible and not let arbitrary decisions get in the way.

The release tag capability is already incredibly useful, I have to thank you again for pointing this out. Indeed I have started using it, and changed my release policy to R-universe first, see for example: https://shikokuchuo.net/mirai/#installation This doesn’t mean I stop publishing on CRAN, but it means I choose when to push out releases and CRAN can catch up as and when it lets me.

**Comment by wlandau** [comment]
Also, we can simplify a lot further: just keep packages.json and have PRs edit it directly. That way the CI is simple and cheap: just check the new and old packages.json, require ‘branch: “*releases”’, and require manual review if the PR causes a name conflict. I will see if I can find time to prototype something this week or next. 

By the way: if we create an r-releases GitHub org, will we have to pay for CI/CD? I forget the pricing on that. An org would allow a team of developers (although maintenance should be super easy if most merges are automatic.)

**Comment by wlandau** [comment]
To be clear: I don't intend to be the next CRAN-like gatekeeper, I just want to pilot this idea to see if it will work. If this is successful, I hope the R Consortium and our working group can find a sustainable governance and maintenance system so any required manual gatekeeping takes very little time for any one person.

**Comment by wlandau** [comment]
>  I am experimenting with this right now, and although some of the basics seem easy to implement, automatic merges might be tough. The pull request at https://github.com/wlandau/test-actions/pull/3 tries to modify its own GitHub Actions YAML file to merge itself unconditionally, which is something users should theoretically not be able to do. For some strange reason, the PR's version of the YAML file runs, not the one in the upstream fork. And the merge is blocked by a 403 error, which makes me think a conditional merge would not have worked anyway. Is there a better way to implement automatic merges when users just want to contribute their packages?

Actually, I think it will work if all the CI/CD happens from a different repo in https://github.com.com/r-releases. Idea: a package author opens a PR in https://github.com/r-releases/r-releases.r-universe.dev, which triggers a job in  https://github.com/r-releases/workflows. The job in `r-releases/workflows` does all the checking and then automatically merges the `r-releases/r-releases.r-universe.dev` PR if checks pass. I think the automatic merge should work because both repos will be part of the same org.

**Comment by wlandau** [comment]
And it should be more secure because all the GitHub Actions workflows are walled behind a repo that will always be manually reviewed. I plan to add high levels of branch protection to `r-releases/workflows`.

**Comment by wlandau** [comment]
> which triggers a job in https://github.com/r-releases/workflows. 

To encapsulate all the automation in `r-releases/workflows`, automatic merging might have to happen as a cron job. Which is fine too.

**Comment by wlandau** [comment]
I have a proof of concept without fancy automation, and I hope to open-source that soon.

**Comment by wlandau** [comment]
Fancy automation is actually going well. It works locally, and am testing it internally. In my current setup, the universe itself has only `packages.json` and GitHub Actions in it are disabled for security because of the pull request automation. (@jeroen, if I disable actions in the repo itself, will that interfere with `r-universe`'s actions?) Then there is a separate tightly-controlled repo which checks the pull requests in the universe and has just enough fine-grained permissions to merge and comment.

**Comment by wlandau** [comment]
The only problem I see is that there may be a lot of merge conflicts when requesting to edit `packages.json`, especially because they are processed in bulk every hour. I am not sure how to handle this other than to alert the user with a comment and an issue label when a merge fails.

**Comment by jeroen** [comment]
> if I disable actions in the repo itself, will that interfere with r-universe's actions?

No, the the actions that run under http://github.com/r-universe are unrelated to the settings in your registry repo. 

Maybe we could consider designing another format for the multiverse that is less sensitive to merge conflicts than `packages.json`, although I am not sure what that would look like. We would probably need a separate file for each package if you want to ensure there are no conflicts.

**Comment by wlandau** [comment]
As another workaround, I wonder if the PR can trigger a webhook immediately, as opposed to letting multiple PRs pile up every hour. I'm not sure how to set this up though. Maybe it will make more sense when I get permission to open-source the current automation scheme I cooked up.

**Comment by wlandau** [comment]
I do like the guardrails of having a single `packages.json`. Currently I have automated checks to make sure the user does not create or move any files, and in the current scheme it seems easier to tell if the user has good intentions.

**Comment by wlandau** [comment]
Actually, I think it would be easy to create an upstream repo with one file per package, then dynamically generate the `packages.json` every hour. And safeguards would actually be easier.

**Comment by wlandau** [comment]
Yeah, one file per package really simplifies things nicely!

**Comment by wlandau** [comment]
Wow, after implementing separate files for separate packages, everything came together so nicely.

How it will work: upstream there will be a repo at https://github.com/r-releases/r-releases where users will contribute packages. For each package,  they just upload a text file, where the name of the file is the package name and the contents of the file is just the URL. Not only does this avoid merge conflicts, it also allows us to check each new package with a lightweight API request instead of a full clone. Even the URL can be checked with the "patch" field of the response. And another job in https://github.com/r-releases/r-releases.r-universe.dev will run every hour to build `packages.json` from these small text files.

**Comment by wlandau** [comment]
And I don't see a hard limit on the number of files in a repo: https://docs.github.com/en/repositories/creating-and-managing-repositories/repository-limits

**Comment by wlandau** [comment]
The prototype is open-sourced at https://github.com/r-releases/r-releases. The automated PR reviews seems to work, but I have temporarily disabled GitHub Actions until we figure out https://github.com/r-releases/r-releases/issues/3. Any advice on getting a bot or an app to own a fine-grained token for a workflow? I'm not sure how to do this.

@Joseph-rickert, if there is room on Monday's agenda in the R Repositories WG, it would be great to share https://github.com/r-releases/r-releases with the team.

**Comment by wlandau** [comment]
> Any advice on getting a bot or an app to own a fine-grained token for a workflow? I'm not sure how to do this.

I think I got a GitHub app going for https://github.com/r-releases/r-releases.r-universe.dev/blob/main/.github/workflows/review_pull_requests.yaml, but we will have to see if it merges https://github.com/r-releases/r-releases/pull/5.

**Comment by wlandau** [comment]
Looks like the app is actually working. I think this prototype is ready to share in the working group.

**Comment by wlandau** [comment]
Confirmed: https://r-releases.r-universe.dev/builds includes the latest release [`crew.cluster`](https://github.com/wlandau/crew.cluster), which was added by https://github.com/r-releases/r-releases/pull/5.

**Comment by wlandau** [comment]
Closing in favor of specific issues and discussions at https://github.com/r-releases/help.
